src/legacy/core_plugins/telemetry/server/fetcher.ts

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import moment from 'moment';
+// @ts-ignore
+import fetch from 'node-fetch';
+import { telemetryCollectionManager } from './collection_manager';
+import { getTelemetryOptIn, getTelemetryUsageFetcher } from './telemetry_config';
+import { getTelemetrySavedObject, updateTelemetrySavedObject } from './telemetry_repository';
+import { REPORT_INTERVAL_MS } from '../common/constants';
+import { getXpackConfigWithDeprecated } from '../common/get_xpack_config_with_deprecated';
+
+export class FetcherTask {
+  private readonly checkDurationMs = 60 * 1000 * 5;
+  private intervalId?: NodeJS.Timeout;
+  private lastReported?: number;
+  private isSending = false;
+  private server: any;
+
+  constructor(server: any) {
+    this.server = server;
+  }
+
+  private getInternalRepository = () => {
+    const { getSavedObjectsRepository } = this.server.savedObjects;
+    const { callWithInternalUser } = this.server.plugins.elasticsearch.getCluster('admin');
+    const internalRepository = getSavedObjectsRepository(callWithInternalUser);
+    return internalRepository;
+  };
+
+  private getCurrentConfigs = async () => {
+    const internalRepository = this.getInternalRepository();
+    const telemetrySavedObject = await getTelemetrySavedObject(internalRepository);
+    const config = this.server.config();
+    const currentKibanaVersion = config.get('pkg.version');
+    const defaultTelemetryUsageFetcher = config.get('telemetry.usageFetcher');
+    const telemetryUrl = getXpackConfigWithDeprecated(config, 'telemetry.url') as string;
+
+    return {
+      telemetryOptIn: getTelemetryOptIn({ currentKibanaVersion, telemetrySavedObject }),
+      telemetryUsageFetcher: getTelemetryUsageFetcher({
+        telemetrySavedObject,
+        defaultTelemetryUsageFetcher,
+      }),
+      telemetryUrl,
+    };
+  };
+
+  private updateLastReported = async () => {
+    const internalRepository = this.getInternalRepository();
+    this.lastReported = Date.now();
+    updateTelemetrySavedObject(internalRepository, {
+      lastReported: this.lastReported,
+    });
+  };
+
+  private checkReportStatus = ({ telemetryOptIn, telemetryUsageFetcher }: any) => {
```

Comment:
nit pick, for sure, but could we rename this to `shouldSendReport` or something like that?
The name `checkReportStatus` forced me to look at how it was used to figure out what the return value actually means.